torch/_utils.py

```diff
@@ -102,44 +104,107 @@ def _accumulate(iterable, fn=lambda x, y: x + y):
+    operating separately.
+
+    In case of sparse tensors, the result will be a tuple of two flat tensors,
+    one for indices and one for values.
+    """
+    if tensors[0].is_sparse:
+        flat_indices = _flatten_tensors([t._indices() for t in tensors])
+        flat_values = _flatten_tensors([t._values() for t in tensors])
+        return flat_indices, flat_values
+    else:
+        if len(tensors) == 1:
+            return tensors[0].contiguous().view(-1)
+        numels = [tensor.numel() for tensor in tensors]
+        size = sum(numels)
+        offset = 0
+        flat = tensors[0].new(size)
+        for tensor, numel in zip(tensors, numels):
+            flat.narrow(0, offset, numel).copy_(tensor, broadcast=False)
+            offset += numel
+        return flat
 
 
 def _unflatten_tensors(flat, tensors):
-    """View a flat buffer using the sizes of tensors"""
+    """View a flat buffer using the sizes of tensors. Assume that tensors are of
+    same type, and that flat is given by _flatten_tensors.
+    """
+    if tensors[0].is_sparse:
+        flat_indices, flat_values = flat
+        indices = _unflatten_tensors(flat_indices, [t._indices() for t in tensors])
+        values = _unflatten_tensors(flat_values, [t._values() for t in tensors])
+        outputs = []
+        for t, i, v in zip(tensors, indices, values):
+            outputs.append(t.new(i, v, t.size()))
+        return tuple(outputs)
+    else:
+        outputs = []
+        offset = 0
+        for tensor in tensors:
+            numel = tensor.numel()
+            outputs.append(flat.narrow(0, offset, numel).view_as(tensor))
+            offset += numel
+        return tuple(outputs)
+
+
+def _reorder_tensors_as(tensors, ordered_tensors):
+    """Assume that tensors are of same order as ordered_tensors within sparse
+    and dense classes. Reordered them to be of same order as ordered_tensors.
+    """
+    sparse_tensors = iter(t for t in tensors if t.is_sparse)
+    dense_tensors = iter(t for t in tensors if not t.is_sparse)
     outputs = []
-    offset = 0
-    for tensor in tensors:
-        numel = tensor.numel()
-        outputs.append(flat.narrow(0, offset, numel).view_as(tensor))
-        offset += numel
+    for t in ordered_tensors:
+        if t.is_sparse:
+            outputs.append(next(sparse_tensors))
+        else:
+            outputs.append(next(dense_tensors))
     return tuple(outputs)
 
 
-def _take_tensors(tensors, size_limit):
-    """Groups tensors into lists of up to size_limit bytes"""
+def _take_tensors(tensors, size_limit, sparse_single_chunk):
+    """Groups tensors into chunks. This generator yeilds a chunk at each call,
+    each containing tensors of up to certain byte limit in total size.
+
+    If sparse_single_chunk=True, yielded tensors are guaranteed to be only of
+    same order within sparse and dense classes, i.e., sparse tensors are of same
+    order with those in input tensors, and same for dense ones.
```

Comment:
Why would you ever want to have this as `False`? Also the name is misleading compared to the docs. The name seems to indicate that if you set it to `False` then sparse tensors will be chunked, however this really means that we enforce a stricter ordering, right?

Reply:
No. The name is intended to indicate that if it is `False` then the sparse tensors will be chunked. That way `broadcast_coalesced` can benefit from it.